File: weather/WeatherModels.swift
//
//  WeatherModels.swift
//  weather
//
//  Created by Matt on 2/4/26.
//

import Foundation

// MARK: - Weather Data Models

struct WeatherData: Codable {
    let latitude: Double
    let longitude: Double
    let current: CurrentWeather
    let hourly: HourlyWeather
    let daily: DailyWeather
}

struct CurrentWeather: Codable {
    let time: String
    let temperature2m: Double
    let apparentTemperature: Double
    let weatherCode: Int
    let windSpeed10m: Double
    let relativeHumidity2m: Int
    
    enum CodingKeys: String, CodingKey {
        case time
        case temperature2m = "temperature_2m"
        case apparentTemperature = "apparent_temperature"
        case weatherCode = "weather_code"
        case windSpeed10m = "wind_speed_10m"
        case relativeHumidity2m = "relative_humidity_2m"
    }
}

struct HourlyWeather: Codable {
    let time: [String]
    let temperature2m: [Double]
    let weatherCode: [Int]
    
    enum CodingKeys: String, CodingKey {
        case time
        case temperature2m = "temperature_2m"
        case weatherCode = "weather_code"
    }
}

struct DailyWeather: Codable {
    let time: [String]
    let weatherCode: [Int]
    let temperature2mMax: [Double]
    let temperature2mMin: [Double]
    let precipitationProbabilityMax: [Int]
    
    enum CodingKeys: String, CodingKey {
        case time
        case weatherCode = "weather_code"
        case temperature2mMax = "temperature_2m_max"
        case temperature2mMin = "temperature_2m_min"
        case precipitationProbabilityMax = "precipitation_probability_max"
    }
}

// MARK: - Weather Condition Helpers

enum WeatherCondition {
    case clearSky
    case partlyCloudy
    case cloudy
    case foggy
    case drizzle
    case rain
    case snow
    case thunderstorm
    case unknown
    
    init(code: Int) {
        switch code {
        case 0:
            self = .clearSky
        case 1, 2:
            self = .partlyCloudy
        case 3:
            self = .cloudy
        case 45, 48:
            self = .foggy
        case 51, 53, 55:
            self = .drizzle
        case 61, 63, 65, 80, 81, 82:
            self = .rain
        case 71, 73, 75, 77, 85, 86:
            self = .snow
        case 95, 96, 99:
            self = .thunderstorm
        default:
            self = .unknown
        }
    }
    
    var description: String {
        switch self {
        case .clearSky: return "Clear Sky"
        case .partlyCloudy: return "Partly Cloudy"
        case .cloudy: return "Cloudy"
        case .foggy: return "Foggy"
        case .drizzle: return "Drizzle"
        case .rain: return "Rain"
        case .snow: return "Snow"
        case .thunderstorm: return "Thunderstorm"
        case .unknown: return "Unknown"
        }
    }
    
    var symbolName: String {
        switch self {
        case .clearSky: return "sun.max.fill"
        case .partlyCloudy: return "cloud.sun.fill"
        case .cloudy: return "cloud.fill"
        case .foggy: return "cloud.fog.fill"
        case .drizzle: return "cloud.drizzle.fill"
        case .rain: return "cloud.rain.fill"
        case .snow: return "cloud.snow.fill"
        case .thunderstorm: return "cloud.bolt.fill"
        case .unknown: return "questionmark.circle.fill"
        }
    }
}

File: android-app/app/src/main/java/com/andernet/weather/data/model/WeatherModels.kt
package com.andernet.weather.data.model

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * Complete weather data response from Open-Meteo API
 */
@JsonClass(generateAdapter = true)
data class WeatherData(
    val latitude: Double,
    val longitude: Double,
    val timezone: String,
    @Json(name = "timezone_abbreviation") val timezoneAbbreviation: String,
    val elevation: Double,
    val current: CurrentWeather?,
    val hourly: HourlyWeather?,
    val daily: DailyWeather?
)

/**
 * Current weather conditions
 */
@JsonClass(generateAdapter = true)
data class CurrentWeather(
    val time: String,
    @Json(name = "temperature_2m") val temperature: Double,
    @Json(name = "apparent_temperature") val apparentTemperature: Double,
    @Json(name = "weather_code") val weatherCode: Int,
    @Json(name = "wind_speed_10m") val windSpeed: Double,
    @Json(name = "wind_direction_10m") val windDirection: Double,
    @Json(name = "wind_gusts_10m") val windGusts: Double?,
    @Json(name = "relative_humidity_2m") val humidity: Int,
    @Json(name = "dew_point_2m") val dewPoint: Double?,
    @Json(name = "pressure_msl") val pressure: Double,
    @Json(name = "cloud_cover") val cloudCover: Int,
    val visibility: Double?,
    @Json(name = "uv_index") val uvIndex: Double?,
    val precipitation: Double,
    @Json(name = "is_day") val isDay: Int
)

/**
 * Hourly forecast data (arrays of 168 hours = 7 days)
 */
@JsonClass(generateAdapter = true)
data class HourlyWeather(
    val time: List<String>,
    @Json(name = "temperature_2m") val temperature: List<Double>,
    @Json(name = "weather_code") val weatherCode: List<Int>,
    @Json(name = "precipitation_probability") val precipitationProbability: List<Int>?,
    @Json(name = "wind_speed_10m") val windSpeed: List<Double>,
    @Json(name = "wind_gusts_10m") val windGusts: List<Double>?,
    @Json(name = "relative_humidity_2m") val humidity: List<Int>,
    @Json(name = "uv_index") val uvIndex: List<Double>?
)

/**
 * Daily forecast data (arrays of 14 days)
 */
@JsonClass(generateAdapter = true)
data class DailyWeather(
    val time: List<String>,
    @Json(name = "weather_code") val weatherCode: List<Int>,
    @Json(name = "temperature_2m_max") val temperatureMax: List<Double>,
    @Json(name = "temperature_2m_min") val temperatureMin: List<Double>,
    @Json(name = "precipitation_probability_max") val precipitationProbabilityMax: List<Int>?,
    val sunrise: List<String>,
    val sunset: List<String>,
    @Json(name = "uv_index_max") val uvIndexMax: List<Double>?,
    @Json(name = "wind_speed_10m_max") val windSpeedMax: List<Double>
)

/**
 * Weather condition types mapped from WMO codes
 */
enum class WeatherCondition(val code: IntRange, val description: String, val iconName: String) {
    CLEAR(0..1, "Clear", "clear_day"),
    PARTLY_CLOUDY(2..2, "Partly Cloudy", "partly_cloudy"),
    CLOUDY(3..3, "Cloudy", "cloudy"),
    OVERCAST(45..48, "Fog", "fog"),
    DRIZZLE(51..57, "Drizzle", "drizzle"),
    RAIN(61..67, "Rain", "rain"),
    SNOW(71..77, "Snow", "snow"),
    RAIN_SHOWERS(80..82, "Rain Showers", "rain"),
    SNOW_SHOWERS(85..86, "Snow Showers", "snow"),
    THUNDERSTORM(95..99, "Thunderstorm", "thunderstorm");

    companion object {
        fun fromCode(code: Int): WeatherCondition {
            return values().find { code in it.code } ?: CLEAR
        }
    }
}

/**
 * Hourly forecast item for UI display
 */
data class HourlyForecastItem(
    val time: String,
    val temperature: Double,
    val weatherCode: Int,
    val precipitationProbability: Int,
    val windSpeed: Double,
    val uvIndex: Double? = null,
    val humidity: Int? = null
)

/**
 * Daily forecast item for UI display
 */
data class DailyForecastItem(
    val date: String,
    val weatherCode: Int,
    val temperatureMax: Double,
    val temperatureMin: Double,
    val precipitationProbability: Int,
    val sunrise: String,
    val sunset: String
)
